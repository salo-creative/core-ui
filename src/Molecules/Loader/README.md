# Loader

Renders various styles of loaders for use in the application. All are styled components and the animations are provided by pure css

---

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Loader from '@salo/core-ui/Molecules/Loader';
```

Implement as follows

```javascript
<Loader display={ true } />
```

**For full prop types and usage see storybook info/knobs**