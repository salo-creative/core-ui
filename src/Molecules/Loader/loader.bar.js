import React from 'react';
import styled, { css, keyframes } from 'styled-components';
import PropTypes from 'prop-types';

const stretchDelay = keyframes`
  0%, 40%, 100% {
    transform: scaleY(0.4);
  }
  20% {
    transform: scaleY(1.0);
  }
`;

const animation = () => css`
  ${ stretchDelay } 1.2s infinite ease-in-out;
`;

// Styling
const StyledLoader = styled.div`
  width: ${ ({ size }) => size }px;
  height: ${ ({ size }) => size }px;
  text-align: center;
  font-size: 10px;
  display: flex;
  justify-content: space-between;
  ${ ({ position, takeover }) => (position === 'absolute' || takeover ? `
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate3d(-50%, -50%, 0);` : 'margin: 0 auto;') }
  > div {
    background-color: ${ ({ appearance, takeover }) => ((appearance === 'dark' && !takeover) || (appearance === 'light' && takeover) ? '#000' : '#fff') };
    height: 100%;
    width: ${ ({ size }) => Math.ceil(size * 0.1333) }px;
    display: flex;
    animation: ${ animation }
  }
  .rect2 {
    animation-delay: -1.1s;
  }
  .rect3 {
    animation-delay: -1.0s;
  }
  .rect4 {
    animation-delay: -0.9s;
  }
  .rect5 {
    animation-delay: -0.8s;
  }
`;

const Bar = (props) => {
  return (
    <StyledLoader { ...props } data-testid='bar-loader'>
      <div className='rect1' />
      <div className='rect2' />
      <div className='rect3' />
      <div className='rect4' />
      <div className='rect5' />
    </StyledLoader>
  );
};

// Default props
Bar.defaultProps = {
  size: 100,
  appearance: 'dark',
  position: 'static'
};

// Prop type check
Bar.propTypes = {
  size: PropTypes.number,
  appearance: PropTypes.oneOf(['dark', 'light']),
  position: PropTypes.oneOf(['static', 'absolute'])
};

export default Bar;