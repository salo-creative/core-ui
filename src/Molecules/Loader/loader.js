import React from 'react';
import PropTypes from 'prop-types';

import Pulse from './loader.pulse';
import Bar from './loader.bar';
import { Background } from './loader.styles';

// component
const Loader = (props) => {
  const { display, loaderProps, takeover, type, appearance } = props;

  const renderTakeover = () => {
    if (takeover && display) {
      return <Background appearance={ appearance } data-testid='loader-takeover' />;
    }
    return null;
  };

  const renderLoader = () => {
    if (!display) return null;
    switch (type) {
      case 'bar': {
        return (
          <Bar
            { ...loaderProps }
            takeover={ takeover }
            appearance={ appearance }
          />
        );
      }
      case 'pulse':
      default: {
        return (
          <Pulse
            { ...loaderProps }
            takeover={ takeover }
            appearance={ appearance }
          />
        );
      }
    }
  };

  return (
    <React.Fragment>
      { renderTakeover() }
      { renderLoader() }
    </React.Fragment>
  );
};

Loader.defaultProps = {
  display: false,
  takeover: false,
  appearance: 'dark',
  type: 'pulse',
  loaderProps: {}
};

Loader.propTypes = {
  display: PropTypes.bool,
  takeover: PropTypes.bool,
  appearance: PropTypes.oneOf(['dark', 'light']),
  type: PropTypes.oneOf(['pulse', 'bar']),
  loaderProps: PropTypes.object
};

export default Loader;