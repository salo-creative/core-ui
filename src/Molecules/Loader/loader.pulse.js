import React from 'react';
import styled, { css, keyframes } from 'styled-components';
import PropTypes from 'prop-types';


const scaleFadeWhite = keyframes`
  0% {
    background-color: rgba(255, 255, 255, 0);
    transform: scale(0);
  }
  5% {
    background-color: rgba(255, 255, 255, 1);
  }
  100% {
    background-color: rgba(255, 255, 255, 0);
    transform: scale(1);
  }
`;

const scaleFadeBlack = keyframes`
  0% {
    background-color: rgba(0, 0, 0, 0);
    transform: scale(0);
  }
  5% {
    background-color: rgba(0, 0, 0, 1);
  }
  100% {
    background-color: rgba(0, 0, 0, 0);
    transform: scale(1);
  }
`;

const animation = props => css`
  ${ (props.appearance === 'light' && !props.takeover) || (props.appearance === 'dark' && props.takeover) ? scaleFadeWhite : scaleFadeBlack } 1.3s 0s linear infinite;
`;

// Styling
const StyledLoader = styled.div`
  box-sizing: border-box;
  position: relative;
  height: ${ props => props.size }px;
  width: ${ props => props.size }px;
  will-change: transform, opacity;
  ${ props => (props.position === 'absolute' || props.takeover ? `
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate3d(-50%, -50%, 0);` : 'margin: 0 auto;') }
  &:before,
  &:after,
  div {
    content: '';
    will-change: transform, opacity;
    border-radius: 100%;
    -webkit-animation-fill-mode: both;
    animation-fill-mode: both;
    position: absolute;
    top: 50%;
    left: 50%;
    margin: -${ props => props.size / 2 }px;
    width: ${ props => props.size }px;
    height: ${ props => props.size }px;
    animation: ${ animation };
  }
  &:before {
    animation-delay: -0.4s;
  }
  &:after {
    animation-delay: -0.2s;
  }
`;

const Pulse = (props) => {
  return <StyledLoader { ...props } data-testid='pulse-loader'><div /></StyledLoader>;
};

// Default props
Pulse.defaultProps = {
  size: 60,
  margin: '0',
  appearance: 'dark',
  position: 'static'
};

// Prop type check
Pulse.propTypes = {
  size: PropTypes.number,
  margin: PropTypes.string,
  appearance: PropTypes.oneOf(['dark', 'light']),
  position: PropTypes.oneOf(['static', 'absolute'])
};

export default Pulse;