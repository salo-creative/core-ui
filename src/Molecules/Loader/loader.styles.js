import styled from 'styled-components';

export const Background = styled.div`
  position: fixed;
  left: 0;
  top: 0;
  height: 100%;
  width: 100%;
  background: ${ ({ appearance }) => (appearance === 'light' ? '#fff' : 'linear-gradient(to bottom, #00222b, #003645)') };
`;