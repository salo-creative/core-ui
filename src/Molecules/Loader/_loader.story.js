import React from 'react';
import {
  withKnobs, boolean, select, object
} from '@storybook/addon-knobs';
import styled from 'styled-components';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Loader } from '../../index';

// README //
import README from './README.md';

// Start of story logic
const Wrapper = styled.div`
  width: 100%;
  min-height: 100vh;
  padding: 0;
`;

export const Basic = () => {
  // KNOBS
  const appearance = select(
    'Appearance',
    {
      Dark: 'dark',
      Light: 'light'
    },
    'dark'
  );
  const type = select(
    'Type',
    {
      Pulse: 'pulse',
      Bar: 'bar'
    },
    'pulse'
  );
  const display = boolean('Display loader', true);
  const takeover = boolean('Display full page takeover', false);
  const loaderProps = object('Loader props', {
    position: 'static',
    size: 100
  });
  return (
    <Wrapper>
      <Loader
        type={ type }
        takeover={ takeover }
        appearance={ appearance }
        display={ display }
        loaderProps={ loaderProps }
      />
    </Wrapper>
  );
};

Basic.story = {
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    est: ['loader'],
    info: {
      propTablesExclude: [Wrapper]
    },
    notes: README
  }
};

export default {
  title: 'Molecules/Loader'
};