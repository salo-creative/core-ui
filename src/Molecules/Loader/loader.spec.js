import React from 'react';
import 'jest-styled-components';

// HELPERS
import { renderWithTheme } from '../../../test';

// COMPONENT
import Loader from './index';

test('Check default render doesnt show Loader and styles', async () => {
  // Render
  const { container } = renderWithTheme(
    <Loader />
  );
  // Assert
  expect(container.firstChild).toBeFalsy();
  expect(container.firstChild).toMatchSnapshot();
});

test('Check default Loader and styles render', async () => {
  // Render
  const { container, getByTestId } = renderWithTheme(
    <Loader
      display={ true }
      data-testid='test-loader'
    />
  );
  const el = getByTestId('pulse-loader');
  let takeover = null;
  try {
    takeover = getByTestId('loader-takeover');
  } catch (e) {
    // Not required just prevents test from throwing
  }
  // Assert
  expect(el).toHaveStyleRule('position', 'relative');
  expect(el).toHaveStyleRule('margin', '0 auto');
  expect(takeover).toBeFalsy();
  expect(container.firstChild).toMatchSnapshot();
});

test('Check light theme Loader and styles render', async () => {
  // Render
  const { container, getByTestId } = renderWithTheme(
    <Loader
      display={ true }
      appearance='light'
      loaderProps={ { position: 'absolute' } }
      data-testid='test-loader'
    />
  );
  const el = getByTestId('pulse-loader');
  // Assert
  expect(el).toHaveStyleRule('position', 'absolute');
  expect(el).toHaveStyleRule('top', '50%');
  expect(el).toHaveStyleRule('left', '50%');
  expect(el).toHaveStyleRule('transform', 'translate3d(-50%,-50%,0)');
  expect(container.firstChild).toMatchSnapshot();
});

test('Check takeover Loader and styles render', async () => {
  // Render
  const { container, getByTestId } = renderWithTheme(
    <Loader
      display={ true }
      takeover={ true }
      data-testid='test-loader'
    />
  );
  const el = getByTestId('pulse-loader');
  const takeover = getByTestId('loader-takeover');
  // Assert
  expect(el).toHaveStyleRule('position', 'absolute');
  expect(el).toHaveStyleRule('top', '50%');
  expect(el).toHaveStyleRule('left', '50%');
  expect(el).toHaveStyleRule('transform', 'translate3d(-50%,-50%,0)');
  expect(takeover).toHaveStyleRule('background', 'linear-gradient(to bottom,#00222b,#003645)');
  expect(container.firstChild).toMatchSnapshot();
});

test('Check takeover light appearance Loader and styles render', async () => {
  // Render
  const { container, getByTestId } = renderWithTheme(
    <Loader
      display={ true }
      takeover={ true }
      appearance='light'
      data-testid='test-loader'
    />
  );
  const takeover = getByTestId('loader-takeover');
  // Assert
  expect(takeover).toHaveStyleRule('background', '#fff');
  expect(container.firstChild).toMatchSnapshot();
});

test('Check bar Loader and styles render', async () => {
  // Render
  const { container, getByTestId } = renderWithTheme(
    <Loader
      display={ true }
      type='bar'
      data-testid='test-loader'
    />
  );
  const el = getByTestId('bar-loader');
  // Assert
  expect(el).toHaveStyleRule('margin', '0 auto');
  expect(container.firstChild).toMatchSnapshot();
});

test('Check light theme bar Loader and styles render', async () => {
  // Render
  const { container, getByTestId } = renderWithTheme(
    <Loader
      display={ true }
      type='bar'
      appearance='light'
      loaderProps={ { position: 'absolute' } }
      data-testid='test-loader'
    />
  );
  const el = getByTestId('bar-loader');
  // Assert
  expect(el).toHaveStyleRule('position', 'absolute');
  expect(el).toHaveStyleRule('top', '50%');
  expect(el).toHaveStyleRule('left', '50%');
  expect(el).toHaveStyleRule('transform', 'translate3d(-50%,-50%,0)');
  expect(container.firstChild).toMatchSnapshot();
});