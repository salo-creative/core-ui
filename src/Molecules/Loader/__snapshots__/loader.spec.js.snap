// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check bar Loader and styles render 1`] = `
.c0 {
  width: 100px;
  height: 100px;
  text-align: center;
  font-size: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 0 auto;
}

.c0 > div {
  background-color: #000;
  height: 100%;
  width: 14px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-animation: eUnCpy 1.2s infinite ease-in-out;
  animation: eUnCpy 1.2s infinite ease-in-out;
}

.c0 .rect2 {
  -webkit-animation-delay: -1.1s;
  animation-delay: -1.1s;
}

.c0 .rect3 {
  -webkit-animation-delay: -1.0s;
  animation-delay: -1.0s;
}

.c0 .rect4 {
  -webkit-animation-delay: -0.9s;
  animation-delay: -0.9s;
}

.c0 .rect5 {
  -webkit-animation-delay: -0.8s;
  animation-delay: -0.8s;
}

<div
  class="c0"
  data-testid="bar-loader"
  size="100"
>
  <div
    class="rect1"
  />
  <div
    class="rect2"
  />
  <div
    class="rect3"
  />
  <div
    class="rect4"
  />
  <div
    class="rect5"
  />
</div>
`;

exports[`Check default Loader and styles render 1`] = `
.c0 {
  box-sizing: border-box;
  position: relative;
  height: 60px;
  width: 60px;
  will-change: transform,opacity;
  margin: 0 auto;
}

.c0:before,
.c0:after,
.c0 div {
  content: '';
  will-change: transform,opacity;
  border-radius: 100%;
  -webkit-animation-fill-mode: both;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
  position: absolute;
  top: 50%;
  left: 50%;
  margin: -30px;
  width: 60px;
  height: 60px;
  -webkit-animation: eyGOfj 1.3s 0s linear infinite;
  animation: eyGOfj 1.3s 0s linear infinite;
}

.c0:before {
  -webkit-animation-delay: -0.4s;
  animation-delay: -0.4s;
}

.c0:after {
  -webkit-animation-delay: -0.2s;
  animation-delay: -0.2s;
}

<div
  class="c0"
  data-testid="pulse-loader"
  size="60"
>
  <div />
</div>
`;

exports[`Check default render doesnt show Loader and styles 1`] = `null`;

exports[`Check light theme Loader and styles render 1`] = `
.c0 {
  box-sizing: border-box;
  position: relative;
  height: 60px;
  width: 60px;
  will-change: transform,opacity;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate3d(-50%,-50%,0);
  -ms-transform: translate3d(-50%,-50%,0);
  transform: translate3d(-50%,-50%,0);
}

.c0:before,
.c0:after,
.c0 div {
  content: '';
  will-change: transform,opacity;
  border-radius: 100%;
  -webkit-animation-fill-mode: both;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
  position: absolute;
  top: 50%;
  left: 50%;
  margin: -30px;
  width: 60px;
  height: 60px;
  -webkit-animation: lkOjTQ 1.3s 0s linear infinite;
  animation: lkOjTQ 1.3s 0s linear infinite;
}

.c0:before {
  -webkit-animation-delay: -0.4s;
  animation-delay: -0.4s;
}

.c0:after {
  -webkit-animation-delay: -0.2s;
  animation-delay: -0.2s;
}

<div
  class="c0"
  data-testid="pulse-loader"
  size="60"
>
  <div />
</div>
`;

exports[`Check light theme bar Loader and styles render 1`] = `
.c0 {
  width: 100px;
  height: 100px;
  text-align: center;
  font-size: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate3d(-50%,-50%,0);
  -ms-transform: translate3d(-50%,-50%,0);
  transform: translate3d(-50%,-50%,0);
}

.c0 > div {
  background-color: #fff;
  height: 100%;
  width: 14px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-animation: eUnCpy 1.2s infinite ease-in-out;
  animation: eUnCpy 1.2s infinite ease-in-out;
}

.c0 .rect2 {
  -webkit-animation-delay: -1.1s;
  animation-delay: -1.1s;
}

.c0 .rect3 {
  -webkit-animation-delay: -1.0s;
  animation-delay: -1.0s;
}

.c0 .rect4 {
  -webkit-animation-delay: -0.9s;
  animation-delay: -0.9s;
}

.c0 .rect5 {
  -webkit-animation-delay: -0.8s;
  animation-delay: -0.8s;
}

<div
  class="c0"
  data-testid="bar-loader"
  size="100"
>
  <div
    class="rect1"
  />
  <div
    class="rect2"
  />
  <div
    class="rect3"
  />
  <div
    class="rect4"
  />
  <div
    class="rect5"
  />
</div>
`;

exports[`Check takeover Loader and styles render 1`] = `
.c0 {
  position: fixed;
  left: 0;
  top: 0;
  height: 100%;
  width: 100%;
  background: linear-gradient(to bottom,#00222b,#003645);
}

<div
  class="c0"
  data-testid="loader-takeover"
/>
`;

exports[`Check takeover light appearance Loader and styles render 1`] = `
.c0 {
  position: fixed;
  left: 0;
  top: 0;
  height: 100%;
  width: 100%;
  background: #fff;
}

<div
  class="c0"
  data-testid="loader-takeover"
/>
`;
